util/fpga/splice.py

```diff
@@ -0,0 +1,381 @@
+
+        Args:
+          infile: str; The input bitstream.
+          outfile: str; Path to the output file.
+        """
+        cmd = [
+            self._opentitantool,
+            '--rcfile=',
+            '--logging=info',
+            'fpga',
+            'update-usr-access',
+            infile,
+            outfile,
+        ]
+        subprocess.check_call(cmd)
+
+    def splice(self, infile, mmis, datas, outfile, swap_nibbles):
+        """Splice multiple files into a bitstream and update USR_ACCESS.
+
+        Args:
+          infile: str; The input bitstream.
+          mmis: List[str]; Paths to MMI memory layout files.
+          datas: List[str]; Paths to data files to splice into the bitstream.
+          outfile: file: The already-opened output file.
+          swap_nibbles: bool; Whether to swap nibbles.
+        Return:
+          bytes: The spliced bitstream data
+        """
+        with tempfile.TemporaryDirectory(prefix='splice-') as tmp:
+            # Create the mem files from the input data.
+            for i, data in enumerate(datas):
+                self.gen_vivado_image(data, f'{tmp}/{i}.mem', swap_nibbles)
+
+            if not infile.endswith('.bit'):
+                os.symlink(infile, f'{tmp}/infile.bit')
+                infile = f'{tmp}/infile.bit'
+
+            # Splice each data file into the bitstream.
+            for i, mmi in enumerate(mmis):
+                tmpbit = f'{tmp}/{i}.bit'
+                self.updatemem(infile, mmi, f'{tmp}/{i}.mem', tmpbit)
+                infile = tmpbit
+
+            # Update the usr_access_value to make this bitstream uniquely identifiable.
+            self.update_usr_access(infile, f'{tmp}/updated.bit')
+
+            # Copy the final bitstream into the cache and return the bitstream content.
+            with open(f'{tmp}/updated.bit', 'rb') as f:
+                content = f.read()
+                if outfile:
+                    outfile.seek(0)
+                    outfile.truncate()
+                    outfile.write(content)
+                    outfile.seek(0)
+                return content
+
+
+def main(args):
+    mmi = args.mmi.split(',')
+    data = args.data.split(',')
+    if len(data) == 0:
+        logging.error('No data files to splice')
+        return 1
+    if len(data) != len(mmi):
+        logging.error(
+            'There must be a 1:1 correspondence between data and mmi files')
+        return 1
+
+    cachedir = os.path.expanduser(args.cache)
+    os.makedirs(os.path.join(cachedir, 'splice'), exist_ok=True)
+
+    tools = Tools(args)
+    key = tools.hash([args.bitstream] + data + mmi)
+
+    # We open and lock the cache's bitstream file so that any simultaneous
+    # attempts to create or access the bitstream file will block here
+    # and allow one process to generate the bitstream.
+    filename = f'splice/{key}.bit'
+    bitstream = open_locked(os.path.join(cachedir, filename), binary=True)
```

Comment:
Is this a blocking call, or will the program return early if unable to acquire the lock?

Reply:
The lock operation (`fcntl.lockf`) is a blocking operation (unless `LOCK_NB` is specified).  Added a comment.